### ISSUE TITLE
MPI validation / regression test (due 8/14)

### ISSUE THREAD

**Issue by gunsch** [department-of-veterans-affairs/va.gov-team#12042]
**Mandate**: OIT leadership mandates that all applications that consume IAM VA MPI ensure they are processing _all_ components of Source ID. No value should be defaulted and ignored.

I've forwarded an email to both of you with more details.

**Comment by alexpappasoddball** [comment]
Is there more context to this ticket?  Seems like the deadline is pretty quick and not sure what exactly has to be done based on the ticket.

**Comment by johnpaulashenfelter** [comment]
From: Podolec, Jeff <Jeff.Podolec@va.gov>
Sent: Monday, August 3, 2020 4:18 PM
To: Podolec, Jeff <Jeff.Podolec@va.gov>
Cc: Reed, Danny L. <Danny.Reed@va.gov>; Kiboti, Patrick (Engility) <Patrick.Kiboti@va.gov>
Subject: Data Call-- Action required by 08/14/20 to validate your system’s consumption of VA MPI services
Importance: High

 

Why are you being notified?  You are identified as someone responsible for a system that consumes VA MPI services.

 

Issue?  To mitigate a possible Patient Safety issue, OIT leadership is mandating that all applications consuming the MPI Search/Retrieve with Corresponding IDs [1305/1306] or GetCorrespondingIDs [1309] services must do regression testing to confirm you are handling the IDTYPE correctly, and if not, make needed changes in your system.  (Details of how to perform the needed regression test are provided below.*)

 

Due date for completing the regression testing and any needed changes in your system: August 14, 2020.

 

Actions required?

Regression test: Explicitly regression test that your system’s handling of IDTYPE is working as your system requires. (Again, details of how to perform the needed regression test are provided below.*)
Notification of compliance: After you complete the regression testing and any needed changes, “reply all” to this email reporting you completed the testing and completed any needed changes in your system. To be clear, even if you complete regression testing and determine no changes are required, still send a reply with that information.
 

Technical questions? Send an email to Reed, Danny L. (Danny.Reed@va.gov) and Kiboti, Patrick (Engility) (Patrick.Kiboti@va.gov).

 

General questions? Send an email to Podolec, Jeff (Jeff.Podolec@va.gov).

 

*Details to perform the needed regression test: 

 

Mandate: OIT leadership mandates that all applications that consume IAM VA MPI ensure they are processing all components of Source ID. No value should be defaulted or ignored. (For more specifics and background review see the attached “SourceID Compliance Issue.docx.”) An email reply indicating compliance must be provided for each application as per the Actions Required section above.

 

How to perform your regression test: Perform your regression test using the following list of ICNs that are in the IAM integrated SQA environment. Again, the test is to ensure your system processes all components according to the business rules for your applications.

·         1013145796V783434

·         1012833271V058378

·         1012644974V092999

·         1013025856V239247

·         1011230043V026194

 

How to interpret a fully-qualified Source ID that is returned by MPI Search/Retrieve with Corresponding IDs [1305/1306] or GetCorrespondingIDs [1309] for the ICNs provided directly above: Consider this example of a fully qualified Source ID for a VHA-correlated patient for which your system should be fully processing all fields (see Note below1):

<id root="2.16.840.1.113883.4.349" extension="7238095^PI^442^USVHA^A"/>

 

In this example,

·         “7238095” is the ID in the Patient File

·         “PI” is the Identifier Type that references “Patient Type”

·         “442” is the Assigning Facility

·         “USVHA” is the Assigning Authority

·         “A” is the IDStatus, meaning “Active.”

 

1Note:  When your system uses 1305/1306 or 1309, you will get all the correlated identifiers for a particular ICN, as shown below. If your system only deals with patients, then you would ensure your system is fully processing only those Source IDs highlighted in green because they have Identifier Type of PI.

            <id root="2.16.840.1.113883.4.349" extension="86401^PI^668^USVHA^A"/>

            <id root="2.16.840.1.113883.4.349" extension="102632^PN^635^USDVA^A"/>

            <id root="2.16.840.1.113883.4.349" extension="281199^PI^554^USVHA^A"/>

             <id root="2.16.840.1.113883.4.349" extension="86401^PI^668^USVHA^A"/>

            <id root="2.16.840.1.113883.4.349" extension="102632^PN^635^USDVA^A"/>

             <id root="2.16.840.1.113883.4.349" extension="7238095^PI^442^USVHA^A"/>

             <id root="2.16.840.1.113883.4.349" extension="37834^PN^575^USDVA^A"/>

 

[SourceID Compliance Issue.pdf](https://github.com/department-of-veterans-affairs/va.gov-team/files/5028353/SourceID.Compliance.Issue.pdf)

**Comment by omgitsbillryan** [comment]
I've read the email a few times over, and these are my takeaways:

> Explicitly regression test that your system’s handling of IDTYPE is working as your system requires.

> all applications that consume IAM VA MPI ensure they are processing all components of Source ID. No value should be defaulted or ignored.

I'm not clear what what it means to "fully process" these fields.

**Comment by annaswims** [comment]
Oof. That document is not an easy read, but here's where we parse the id's and it looks like we explicitly look for either `PI` or `PN` in our regex but not both interchangeably.

https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/mvi/responses/id_parser.rb#L28

**Comment by jholton** [comment]
> what it means to "fully process" these fields

They mean consider every component of the IDs.  That is, each value between each carrot.

**Comment by jholton** [comment]
Here is a restatement of the issue that I wrote up:

The problem:
> (Before) 6/2/2020 VA MPI did not share a common Assigning Facility across the different type of correlated records until we started enumerating the VA VistA NEW PERSON file.

As a result of “enumerating the VA VistA NEW PERSON file”, VA MPI  introduced a new VistA correlation ID for “persons” that shared the same Assigning facility as “patients”. 
These new correlation IDs are now included in the responses of person queries to MPI.

The issue became apparent when clinical applications (MHV and CPRS) started reporting clinical data belonging to the wrong patient being pulled into views and clinical notes that are not for the selected patient.
It was also reported that CCDAs had PII for the other patients showing as well. 

In the case of these clinical apps, they are/were interpreting a VistA user/person (PN) correlation as if it was a VistA Patient (PI) correlation and pulling the clinical data for the ID associated with their user record as if it was their patient record in that VistA instance.

The root cause of the issue is with how the consumer apps of VA MPI are erroneously evaluating the new correlated identifiers that correspond to the VA VistA NEW _PERSON_ file (mentioned above) being returned from VA MPI. They were being treated the same as the _PATIENT_ correlation IDs.

A correlation ID follows this pattern:
\<Unique_ID^ID_Type^Assigning_Facility^Assigning_Authority^ID_Status\/\>

A VistA _PERSON_ correlation ID looks like this:
\<id root=“2.16.840.1.113883.4.349" extension=“102632^**PN**^635^USDVA^A”\/>

A VistA _PATIENT_ correlation ID looks like this:
\<id root=“2.16.840.1.113883.4.349" extension=“86401^**PI**^668^USVHA^A”\/\>

**Comment by jholton** [comment]
And here is my response in regards to va.gov's use of these specific correlation IDs:

VA.gov parses correlation IDs from MPI requests in the `id_parser.rb` class.
For a given user, we do [extract facility IDs](https://github.com/department-of-veterans-affairs/vets-api/blob/fce99d83787a0bd2d2de7860e2e489436d40ecda/lib/mvi/responses/id_parser.rb#L57) ("Assigning Facility").
For this we only consider correlation IDs with an ID Type of `PI` and an Assigning Authority of `USVHA`.

This collection of facility IDs is the only place we consider VistA related correlation IDs.
We do not parse out any patient (or person) unique identifiers for any purpose, including retrieving data from VistA-related systems .

tl:dr: We don’t use any VistA ID for any purpose whatsoever.

**Comment by alexpappasoddball** [comment]
Moving this issue over to the identity team.  @christine-dillman double check my labels.

Also removing myself and assigning to Johnny with JP for coverage
